File: serve.js
#!/usr/bin/env node
/**
 * Simple HTTP server for serving the Ricochet game locally using Node.js.
 * This solves the ES6 module CORS issues when running on Linux.
 *
 * Usage:
 *     node serve.js [port]
 *     ./serve.js [port]
 *
 * Default port is 8000
 *
 * Prerequisites:
 *     - Node.js installed
 *     - No additional dependencies required (uses built-in modules)
 */

const http = require('http');
const fs = require('fs');
const path = require('path');
const { spawn } = require('child_process');

// Configuration
const DEFAULT_PORT = 8000;
const MIME_TYPES = {
    '.html': 'text/html',
    '.js': 'application/javascript',
    '.css': 'text/css',
    '.json': 'application/json',
    '.png': 'image/png',
    '.jpg': 'image/jpeg',
    '.jpeg': 'image/jpeg',
    '.gif': 'image/gif',
    '.svg': 'image/svg+xml',
    '.ico': 'image/x-icon',
    '.mp3': 'audio/mpeg',
    '.wav': 'audio/wav',
    '.ogg': 'audio/ogg',
    '.txt': 'text/plain',
    '.md': 'text/markdown'
};

function getMimeType(filePath) {
    const ext = path.extname(filePath).toLowerCase();
    return MIME_TYPES[ext] || 'application/octet-stream';
}

function serveFile(req, res, filePath) {
    fs.readFile(filePath, (err, data) => {
        if (err) {
            if (err.code === 'ENOENT') {
                res.writeHead(404, { 'Content-Type': 'text/html' });
                res.end(`
                    <html>
                        <head><title>404 Not Found</title></head>
                        <body>
                            <h1>404 - File Not Found</h1>
                            <p>The requested file <code>${req.url}</code> was not found.</p>
                            <p><a href="/">Back to index</a></p>
                        </body>
                    </html>
                `);
            } else {
                res.writeHead(500, { 'Content-Type': 'text/plain' });
                res.end(`Server Error: ${err.message}`);
            }
            return;
        }

        const mimeType = getMimeType(filePath);
        res.writeHead(200, {
            'Content-Type': mimeType,
            'Access-Control-Allow-Origin': '*',
            'Access-Control-Allow-Methods': 'GET, POST, OPTIONS',
            'Access-Control-Allow-Headers': 'Content-Type',
            'Cache-Control': 'no-cache' // Disable caching for development
        });
        res.end(data);
    });
}

function createServer(port) {
    const server = http.createServer((req, res) => {
        // Handle CORS preflight requests
        if (req.method === 'OPTIONS') {
            res.writeHead(200, {
                'Access-Control-Allow-Origin': '*',
                'Access-Control-Allow-Methods': 'GET, POST, OPTIONS',
                'Access-Control-Allow-Headers': 'Content-Type'
            });
            res.end();
            return;
        }

        // Parse URL
        let urlPath = req.url;
        if (urlPath === '/') {
            urlPath = '/index-modular.html'; // Default to modular version
        }

        // Security: prevent directory traversal
        if (urlPath.includes('..')) {
            res.writeHead(403, { 'Content-Type': 'text/plain' });
            res.end('Forbidden: Directory traversal not allowed');
            return;
        }

        // Convert URL path to file path
        const filePath = path.join(__dirname, urlPath);

        // Check if file exists and is within the project directory
        const projectDir = path.resolve(__dirname);
        const resolvedPath = path.resolve(filePath);

        if (!resolvedPath.startsWith(projectDir)) {
            res.writeHead(403, { 'Content-Type': 'text/plain' });
            res.end('Forbidden: Access outside project directory not allowed');
            return;
        }

        // Serve the file
        serveFile(req, res, filePath);
    });

    return server;
}

function openBrowser(url) {
    const commands = {
        'linux': 'xdg-open',
        'darwin': 'open',
        'win32': 'start'
    };

    const command = commands[process.platform];
    if (command) {
        const child = spawn(command, [url], {
            detached: true,
            stdio: 'ignore'
        });
        child.unref();
        return true;
    }
    return false;
}

function main() {
    // Get port from command line or use default
    let port = DEFAULT_PORT;
    if (process.argv.length > 2) {
        const portArg = parseInt(process.argv[2]);
        if (isNaN(portArg) || portArg < 1 || portArg > 65535) {
            console.error(`❌ Error: Invalid port number '${process.argv[2]}'. Using default port ${DEFAULT_PORT}.`);
        } else {
            port = portArg;
        }
    }

    // Create and start server
    const server = createServer(port);

    server.listen(port, 'localhost', () => {
        console.log('🎮 Ricochet Game Server (Node.js)');
        console.log(`📁 Serving directory: ${__dirname}`);
        console.log(`🌐 Server running at: http://localhost:${port}`);
        console.log('📄 Game URLs:');
        console.log(`   • Original:     http://localhost:${port}/index.html`);
        console.log(`   • Modular:      http://localhost:${port}/index-modular.html`);
        console.log(`   • Level Editor: http://localhost:${port}/level-parser-demo.html`);
        console.log('⚡ Press Ctrl+C to stop the server');
        console.log();

        // Auto-open browser after a short delay
        setTimeout(() => {
            const url = `http://localhost:${port}/index-modular.html`;
            if (openBrowser(url)) {
                console.log('🚀 Opened game in default browser');
            } else {
                console.log('⚠️  Could not auto-open browser');
                console.log('📝 Manually open the URL above in your browser');
            }
        }, 1000);
    });

    server.on('error', (err) => {
        if (err.code === 'EADDRINUSE') {
            console.error(`❌ Error: Port ${port} is already in use`);
            console.error(`💡 Try a different port: node serve.js ${port + 1}`);
        } else {
            console.error(`❌ Error starting server: ${err.message}`);
        }
        process.exit(1);
    });

    // Handle graceful shutdown
    process.on('SIGINT', () => {
        console.log('\n🛑 Server stopped by user');
        server.close(() => {
            process.exit(0);
        });
    });

    process.on('SIGTERM', () => {
        console.log('\n🛑 Server terminated');
        server.close(() => {
            process.exit(0);
        });
    });
}

// Make the script executable
if (require.main === module) {
    main();
}

module.exports = { createServer, getMimeType };

File: serve.py
#!/usr/bin/env python3
"""
Simple HTTP server for serving the Ricochet game locally.
This solves the ES6 module CORS issues when running on Linux.

Usage:
    python3 serve.py [port]
    ./serve.py [port]

Default port is 8000
"""

import http.server
import socketserver
import webbrowser
import sys
import os
import threading
import time

def main():
    # Get port from command line or use default
    port = 8000
    if len(sys.argv) > 1:
        try:
            port = int(sys.argv[1])
        except ValueError:
            print(f"Error: Invalid port number '{sys.argv[1]}'. Using default port 8000.")

    # Change to the directory containing this script
    script_dir = os.path.dirname(os.path.abspath(__file__))
    os.chdir(script_dir)

    # Create server
    handler = http.server.SimpleHTTPRequestHandler

    # Add CORS headers for local development
    class CORSHTTPRequestHandler(handler):
        def end_headers(self):
            self.send_header('Access-Control-Allow-Origin', '*')
            self.send_header('Access-Control-Allow-Methods', 'GET, POST, OPTIONS')
            self.send_header('Access-Control-Allow-Headers', 'Content-Type')
            super().end_headers()

    try:
        with socketserver.TCPServer(("", port), CORSHTTPRequestHandler) as httpd:
            print(f"🎮 Ricochet Game Server")
            print(f"📁 Serving directory: {script_dir}")
            print(f"🌐 Server running at: http://localhost:{port}")
            print(f"📄 Game URLs:")
            print(f"   • Original:     http://localhost:{port}/index.html")
            print(f"   • Modular:      http://localhost:{port}/index-modular.html")
            print(f"   • Level Editor: http://localhost:{port}/level-parser-demo.html")
            print(f"⚡ Press Ctrl+C to stop the server")
            print()

            # Auto-open browser after a short delay
            def open_browser():
                time.sleep(1)
                try:
                    webbrowser.open(f'http://localhost:{port}/index-modular.html')
                    print("🚀 Opened game in default browser")
                except Exception as e:
                    print(f"⚠️  Could not auto-open browser: {e}")
                    print("📝 Manually open the URL above in your browser")

            browser_thread = threading.Thread(target=open_browser)
            browser_thread.daemon = True
            browser_thread.start()

            # Start server
            httpd.serve_forever()

    except KeyboardInterrupt:
        print("\n🛑 Server stopped by user")
        sys.exit(0)
    except OSError as e:
        if e.errno == 98:  # Address already in use
            print(f"❌ Error: Port {port} is already in use")
            print(f"💡 Try a different port: python3 serve.py {port + 1}")
        else:
            print(f"❌ Error starting server: {e}")
        sys.exit(1)
    except Exception as e:
        print(f"❌ Unexpected error: {e}")
        sys.exit(1)

if __name__ == "__main__":
    main()
